Core: physical backing for flexible and pooled memory allocations

StevenMiller123 opened the pull request:
This PR handles partial unmaps of flexible memory. I've created a separate flexible memory map used to track flexible "physical" allocations, calculate free physical addresses to map, and implemented logic in MapMemory and UnmapBytesFromEntry for properly updating this flexible map.

I also added a fix for the isDevKit logic, the backing size wasn't properly updated for the increased direct memory size of isDevKit, which could cause weird issues with direct memory allocations that use the extra space.

This improves Uncharted 4: A Thief’s End™ (CUSA00341), The Last of Us™ Part II (CUSA07820), and FINAL FANTASY XV (CUSA01633) on Windows devices.

Commit:
Fix isDevKit

Previously, isDevKit could increase the physical memory used above the length we reserve in the backing file.

Commit:
Physical backing for flexible allocations

I took the simple approach here, creating a separate map for flexible allocations and pretty much just copying over the logic used in the direct memory map.

Commit:
Various fixups

Commit:
Fix mistake #1

Commit:
Assert + clang

StevenMiller123 committed:
Fix NameVirtualRange

Commit:
Fix 2

StevenMiller123 committed:
Clang

Commit:
Fix CanMergeWith

Validate physical base for flexible mappings

Commit:
Clang

Commit:
Physical backing for pooled memory

Commit:
Allow VMA splitting in NameVirtualRange

This should be safe, since with the changes in this PR, the only issues that come from discrepancies between address space and vma_map are issues related to vmas being larger than address space mappings. NameVirtualRange will only ever shrink VMAs by naming part of one.

Commit:
Fix

Niram7777 commented on src/core/memory.cpp:
```diff
@@ -41,8 +41,11 @@ void MemoryManager::SetupMemoryRegions(u64 flexible_size, bool use_extended_mem1
-        total_size += is_neo ? 2_GB : 768_MB;
         LOG_WARNING(Kernel_Vmm,
                     "Config::isDevKitConsole is enabled! Added additional {:s} of direct memory.",
                     is_neo ? "2 GB" : "768 MB");
```
more readable/consistent ?
SCE_KERNEL_TOTAL_MEM_DEV = 6016_MB
SCE_KERNEL_TOTAL_MEM_PRO_DEV = 7936_MB

maybe `// Assuming 2gb is neo for now,` should go in memory.h next do MEM_*DEV constants

maybe log
`{:d} MB of direct memory.",(total_size - old_size) ;`
intead of
`is_neo ? "2 GB" : "768 MB")`

Commit:
Revert NameVirtualRange changes

Seems like it doesn't play nice for Windows

Commit:
Clean up isDevKit logic

We already log both isNeo and isDevKit in Emulator::Run, so the additional logging in MemoryManager::SetupMemoryRegions isn't really necessary.

I've also added a separate constant for non-pro devkit memory, as suggested.

Finally I've changed a couple constants to use the ORBIS prefix we generally follow here, instead of the SCE prefix.

Commit:
Erase flexible memory contents from physical memory on unmap

Flexible memory should not be preserved on unmap, so erase flexible contents from the physical backing when unmapping.

Commit:
Expand flexible memory map

Some games will end up fragmenting the physical backing space used for flexible memory. To reduce the frequency of this happening under normal circumstances, allocate the entirety of the remaining physical backing to the flexible memory map.

This is effectively a workaround to the problem, but at the moment I think this should suffice.

Commit:
Clang
